modules/drivers/audio/speaker_component.cc

```diff
@@ -0,0 +1,105 @@
+ * limitations under the License.
+ *****************************************************************************/
+
+#include "modules/drivers/audio/speaker_component.h"
+#include <google/protobuf/message.h>
+
+namespace apollo {
+namespace drivers {
+namespace audio {
+
+bool SpeakerComponent::Init() {
+  speaker_config_ = std::make_shared<SpeakerConfig>();
+  if (!apollo::cyber::common::GetProtoFromFile(config_file_path_,
+                                               speaker_config_.get())) {
+    return false;
+  }
+  AINFO << "Speaker config: " << speaker_config_->DebugString();
+
+  // new speaker device
+  speaker_device_.reset(new Respeaker());
+  speaker_device_->init(speaker_config_);
+
+  // dump config, calculate size and reserve buffer for audio
+  // chunk_: number of frames per chunk; chunk_size_: number of bytes per chunk
+  n_channels_ = speaker_config_->channel_type_size();
+  chunk_ = speaker_config_->chunk();
+  total_frames_ =
+      speaker_config_->sample_rate() * speaker_config_->record_seconds();
+  chunk_size_ = speaker_device_->stream_->get_chunk_size(chunk_);
+
+  buffer_size_ = speaker_device_->stream_->get_chunk_size(total_frames_);
+  buffer_ = reinterpret_cast<char *>(malloc(buffer_size_));
+  if (buffer_ == nullptr) {
+    AERROR << "System calloc memory error, size:" << buffer_size_;
+    return false;
+  }
+
+  // assemble AudioData -- fill speaker_config, allocate memory for channel_data
+  audio_data_.reset(new AudioData());
+  std::string config;
+  speaker_config_->SerializeToString(&config);
+  audio_data_->mutable_speaker_config()->ParseFromString(config);
+
+  ChannelData *channel_data = nullptr;
+  int channel_size = buffer_size_ / n_channels_;
+  for (int i = 0; i < n_channels_; ++i) {
+    channel_data = audio_data_->add_channel_data();
+    channel_data->set_channel_type(speaker_config_->channel_type(i));
+    channel_data->set_size(channel_size);
+    channel_data->mutable_data()->resize(channel_size);
+    channel_data_ptr_.push_back(channel_data->mutable_data());
+  }
+
+  writer_ = node_->CreateWriter<AudioData>(speaker_config_->channel_name());
+  async_result_ = cyber::Async(&SpeakerComponent::run, this);
+  return true;
+}
+
+void SpeakerComponent::run() {
+  char *pos = nullptr;
+  int i;
+  while (!cyber::IsShutdown()) {
+    AINFO << "Start recording";
+    pos = buffer_;
+    for (i = 1; chunk_ * i <= total_frames_; ++i) {
+      speaker_device_->stream_->read_stream(chunk_, pos);
+      pos += chunk_size_;
+    }
+    if (chunk_ * --i < total_frames_)
+      speaker_device_->stream_->read_stream(total_frames_ - chunk_ * i, pos);
+    AINFO << "End recording";
+
+    for (int buff_i = 0, i = 0; buff_i < buffer_size_; i += 2) {
+      for (int channel_i = 0; channel_i < n_channels_; ++channel_i) {
+        (*channel_data_ptr_[channel_i])[i] = buffer_[buff_i++];
+        (*channel_data_ptr_[channel_i])[i + 1] = buffer_[buff_i++];
+      }
+    }
```

Comment:
fulfill header before publish the data.

Reply:
I use `grep -nr 'mutable_head()` under `modules/drivers` to see how other drivers do to header and mimic them, adding the following fields to header
1. `frame_id`: I guess it's the type of config for the hardware, so I add a string in speaker config as "respeaker" and set it to header
2. `sequence_num`: an int incremented by one per message
3. `timestamp_sec`: current time -- `mutable_header()->set_timestamp_sec(cyber::Time().Now().ToSecond());`

If there is any field missed in header please comment here and I will make further changes.

Reply:
> I use `grep -nr 'mutable_head()` under `modules/drivers` to see how other drivers do to header and mimic them, adding the following fields to header
> 
> 1. `frame_id`: I guess it's the type of config for the hardware, so I add a string in speaker config as "respeaker" and set it to header
> 2. `sequence_num`: an int incremented by one per message
> 3. `timestamp_sec`: current time -- `mutable_header()->set_timestamp_sec(cyber::Time().Now().ToSecond());`
> 
> If there is any field missed in header please comment here and I will make further changes.

Yes, those are the critical three fields.

Reply:
Now I change to use `Fill header` in `modules/common` for code reuse. If there's any problems or concerns, please comment here.